src/projection.jl

```diff
@@ -135,13 +135,41 @@ ProjectTo(::Real) = ProjectTo{Real}()
 ProjectTo(::Complex) = ProjectTo{Complex}()
 ProjectTo(::Number) = ProjectTo{Number}()
 for T in (Float16, Float32, Float64, ComplexF16, ComplexF32, ComplexF64)
-    # Preserve low-precision floats as accidental promotion is a common perforance bug
+    # Preserve low-precision floats as accidental promotion is a common performance bug
     @eval ProjectTo(::$T) = ProjectTo{$T}()
 end
 ProjectTo(x::Integer) = ProjectTo(float(x))
 ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
-(::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
-(::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
+
+# Preserve low-precision floats as accidental promotion is a common performance bug
+# In these cases we can just `convert` as we know we are dealing with plain and simple types
+(::ProjectTo{T})(dx::AbstractFloat) where T<:AbstractFloat = convert(T, dx)
+(::ProjectTo{T})(dx::Integer) where T<:AbstractFloat = convert(T, dx)  #needed to avoid ambiguity
+
+# We asked for a number/real and they gave use one. We did ask for a particular concrete
+# type, but that is just for the preserving low precision floats, which is handled above.
+# Any Number/Real actually occupies the same subspace, so we can trust them.
+# In particular, this makes weirder Real subtypes that are not simply the values like
+# ForwardDiff.Dual and Symbolics.Sym work, because we stay out of their way.
+(::ProjectTo{<:Number})(dx::Number) = dx 
+# If you remove the above julia sometimes can't find the (::ProjectTo{T})(::T) for complex T
+# Seems like it might be a julia bug?
+
+(project::ProjectTo{<:Real})(dx::Complex) = project(real(dx))
+
+# Complex 
+
+# Preserve low-precision floats as accidental promotion is a common performance bug
+# In these cases we can just `convert` as we know we are dealing with plain and simple types
+(::ProjectTo{T})(dx::Complex{<:AbstractFloat}) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+(::ProjectTo{T})(dx::AbstractFloat) where {T<:Complex{<:AbstractFloat}} = convert(T, dx)
+
+# For  on-AbstractFloat other types pass though to project each component
+function (::ProjectTo{<:Complex{T}})(dx::Complex) where T
+    project = ProjectTo(zero(T))
+    return Complex(project(real(dx)), project(imag(dx)))
```

Comment:
What's this aimed at? The real types we preserve, we also already preserve their complex versions. Maybe recursing into Complex can wait until someone has a stranger real type which needs it?

Reply:
I thought we needed it for ` ProjectTo(1.0 + 1im)(Complex(Dual(1.0, 2.0), Dual(1.0, 2.0)))`
But that gets caught by `(::ProjectTo{<:Number})(dx::Number) = dx`

We would need it if we wanted `ProjectTo(1.0 + 1im)(Complex(Dual(1, 2), Dual(1, 2)))` to return a `Complex{Dual{Float64}}` 
rather than a `Complex{Dual{Int}}` which it currently does.
but that is not needed right now, so I guess we can add it back later